File: android-sample-java/app/src/main/java/io/logto/demo4j/viewmodel/LogtoViewModelFactory.java
package io.logto.demo4j.viewmodel;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class LogtoViewModelFactory implements ViewModelProvider.Factory {

    private final Application application;

    public LogtoViewModelFactory(@NonNull Application application) {
        this.application = application;
    }

    @NonNull
    @Override
    @SuppressWarnings ("unchecked")
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(LogtoViewModel.class)) {
            return (T) new LogtoViewModel(application);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: android-sample-kotlin/app/src/main/kotlin/io/logto/android/sample4k/viewmodel/LogtoViewModelFactory.kt
package io.logto.android.sample4k.viewmodel

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class LogtoViewModelFactory(private val application: Application) : ViewModelProvider.Factory {
    @Suppress("unchecked_cast")
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(LogtoViewModel::class.java)) {
            return LogtoViewModel(application) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
